x/dsmr/storage.go

```diff
@@ -326,6 +378,46 @@ func (s *ChunkStorage[T]) GetChunkBytes(expiry int64, chunkID ids.ID) ([]byte, e
 	return chunkBytes, nil
 }
 
+func (s *ChunkStorage[T]) CheckRateLimit(chunk Chunk[T]) error {
+	producer := chunk.Producer
+	expiry := chunk.Expiry
+	weight := uint64(len(chunk.bytes))
+	producerChunks := s.pendingChunksProducers[producer]
+	if len(producerChunks) == 0 {
+		return nil
+	}
+	window := s.ruleFactory.GetRules(expiry).GetChunkRateLimitingWindow()
+	weightLimit := s.ruleFactory.GetRules(expiry).GetAccumulatedChunkWeightWindowLimit()
+
+	// remove all the producer chunks prior then [expiry - window]
+	priorChunksIdx := sort.Search(len(producerChunks), func(i int) bool {
+		return producerChunks[i].Chunk.Expiry > expiry-window
+	})
+	if priorChunksIdx != len(producerChunks) {
+		producerChunks = producerChunks[priorChunksIdx:]
+	}
+
+	// remove all the producer chunks after [expiry + window]
+	postChunksIdx := sort.Search(len(producerChunks), func(i int) bool {
+		return producerChunks[i].Chunk.Expiry > expiry+window
+	})
+	if postChunksIdx != len(producerChunks) {
+		producerChunks = producerChunks[:postChunksIdx]
+	}
+
+	// we're left only with the chunks that are in the range of [expiry-window,expiry+window]
+	for i := 0; i < len(producerChunks); i++ {
+		accumulatedWeight := len(producerChunks[i].Chunk.bytes)
+		for j := i + 1; j < len(producerChunks) && producerChunks[i].Chunk.Expiry+window > producerChunks[j].Chunk.Expiry; j++ {
+			accumulatedWeight += len(producerChunks[j].Chunk.bytes)
+		}
+		if uint64(accumulatedWeight)+weight > weightLimit {
+			return ErrChunkRateLimitSurpassed
+		}
+	}
+	return nil
```

Comment:
Is it necessary to compute these on demand?

Reply:
see my response above..